evm/src/NttManager/NttManagerWithPerChainTransceivers.sol

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: Apache 2
+pragma solidity >=0.8.8 <0.9.0;
+
+import "./NttManagerNoRateLimiting.sol";
+
+/// @title  NttManagerWithPerChainTransceivers
+/// @author Wormhole Project Contributors.
+/// @notice The NttManagerWithPerChainTransceivers abstract contract is an implementation of
+///         NttManager that allows configuring different transceivers and thresholds
+///         for each chain. You can configure a different set of send and receive
+///         transceivers for each chain, and if you don't specifically enable any
+///         transceivers for a chain, then all transceivers will be used for it.
+///
+/// @dev    All of the developer notes from `NttManager` apply here.
+contract NttManagerWithPerChainTransceivers is NttManagerNoRateLimiting {
+    constructor(
+        address _token,
+        Mode _mode,
+        uint16 _chainId
+    ) NttManagerNoRateLimiting(_token, _mode, _chainId) {}
+
+    // ==================== Storage slots ==========================================================
+
+    bytes32 private constant SEND_TRANSCEIVER_BITMAP_SLOT =
+        bytes32(uint256(keccak256("nttpct.sendTransceiverBitmap")) - 1);
+
+    bytes32 private constant RECV_TRANSCEIVER_BITMAP_SLOT =
+        bytes32(uint256(keccak256("nttpct.recvTransceiverBitmap")) - 1);
+
+    // ==================== Override / implementation of transceiver stuff =========================
+
+    /// @inheritdoc IManagerBase
+    function enableSendTransceiverForChain(
+        address transceiver,
+        uint16 forChainId
+    ) external override(ManagerBase, IManagerBase) onlyOwner {
+        _enableTransceiverForChain(transceiver, forChainId, SEND_TRANSCEIVER_BITMAP_SLOT);
+        emit SendTransceiverEnabledForChain(transceiver, forChainId);
+    }
+
+    /// @inheritdoc IManagerBase
+    function enableRecvTransceiverForChain(
+        address transceiver,
+        uint16 forChainId
+    ) external override(ManagerBase, IManagerBase) onlyOwner {
+        _enableTransceiverForChain(transceiver, forChainId, RECV_TRANSCEIVER_BITMAP_SLOT);
+
+        // Make sure the threshold is set to something.
+        if (getPerChainThreshold(forChainId) == 0) {
+            this.setPerChainThreshold(forChainId, 1);
+        }
+
+        emit RecvTransceiverEnabledForChain(transceiver, forChainId);
+    }
+
+    function _enableTransceiverForChain(
+        address transceiver,
+        uint16 forChainId,
+        bytes32 tag
+    ) private {
+        if (transceiver == address(0)) {
+            revert InvalidTransceiverZeroAddress();
+        }
+
+        TransceiverInfo storage transceiverInfo = _getTransceiverInfosStorage()[transceiver];
```

Comment:
We have to be very careful about adding functionality to the transceiver registry because the `NttManager` contract is running low on code space. That's why `NttManagerWithPerChainTransceiver` inherits from the no-rate-limiting one rather than NttManager.

Reply:
Oh yeah, you're totally right about that. I'm not a big fan of adding more to the contracts. It feels a bit confusing that the logic now contains both per-chain transceivers, however, there are some functions input params and return values don't have a `chainId`.

Reply:
Please take another look since this has all changed.